File: sdk/typescript/test/client.test.js
import assert from "node:assert/strict";
import test from "node:test";
import { GideonDBAPIError, GideonDBClient, GideonDBTransportError } from "../src/index.js";

function queuedFetch(responses, requests = []) {
  return async (url, options) => {
    requests.push({ url, options });
    const response = responses.shift();
    if (response instanceof Error) throw response;
    return response;
  };
}

function jsonResponse(value, status = 200) {
  return new Response(status === 204 ? null : JSON.stringify(value), {
    status,
    headers: { "Content-Type": "application/json" },
  });
}

test("authenticated lifecycle requests encode path and namespace", async () => {
  const requests = [];
  const client = new GideonDBClient("https://db.example/", {
    apiKey: "secret",
    fetch: queuedFetch([
      jsonResponse({ status: "ok" }),
      jsonResponse({ name: "docs", dimension: 2, metric: "dot", shard_count: 2 }, 201),
      jsonResponse({ id: "a/b", vector: [1, 0], version: 1 }),
      jsonResponse({ id: "a/b", vector: [1, 0], version: 1 }),
      jsonResponse({ results: [{ id: "a/b", score: 1 }] }),
      jsonResponse(null, 204),
    ], requests),
  });
  await client.health();
  await client.createCollection({ name: "docs", dimension: 2, metric: "dot", shard_count: 2 });
  await client.upsert("docs", { id: "a/b", vector: [1, 0] });
  assert.equal((await client.get("docs", "a/b", { namespace: "tenant one" })).version, 1);
  assert.equal((await client.search("docs", { vector: [1, 0], topK: 1 }))[0].id, "a/b");
  await client.delete("docs", "a/b");
  assert.match(requests[3].url, /a%2Fb\?namespace=tenant\+one$/);
  assert.equal(requests[3].options.headers.Authorization, "Bearer secret");
});

test("distributed multi-status and API errors remain explicit", async () => {
  const client = new GideonDBClient("http://db.example", {
    fetch: queuedFetch([
      jsonResponse({ results: [], partial: true, failures: [{ shard_id: 1 }], metadata_epoch: 2, authoritative_placement: true }),
      jsonResponse({ outcomes: [{ shard_id: 0, status: "unknown" }], partial: true, metadata_epoch: 2, authoritative_placement: true }, 207),
      jsonResponse({ code: "not_found", message: "missing" }, 404),
    ]),
  });
  assert.equal((await client.distributedSearch("docs", { vector: [1, 0], topK: 3, allowPartial: true })).partial, true);
  assert.equal((await client.distributedBatchUpsert("docs", [{ id: "one", vector: [1, 0] }], { acknowledgement: "all" })).outcomes[0].status, "unknown");
  await assert.rejects(client.describeCollection("missing"), error => error instanceof GideonDBAPIError && error.statusCode === 404 && error.code === "not_found");
});

test("origin, timeout, cancellation, and response bounds fail safely", async () => {
  assert.throws(() => new GideonDBClient("ftp://db.example"), TypeError);
  assert.throws(() => new GideonDBClient("https://db.example/path"), TypeError);
  assert.throws(() => new GideonDBClient("https://db.example", { timeoutMs: 0 }), TypeError);
  const controller = new AbortController();
  controller.abort();
  const cancelled = new GideonDBClient("https://db.example", { fetch: async (_url, options) => { options.signal.throwIfAborted(); } });
  await assert.rejects(cancelled.health({ signal: controller.signal }), GideonDBTransportError);
  const oversized = new GideonDBClient("https://db.example", { fetch: queuedFetch([new Response(new Uint8Array((16 << 20) + 1))]) });
  await assert.rejects(oversized.health(), GideonDBTransportError);
});

test("scroll preserves opaque cursors and explicit vector inclusion", async () => {
  const requests = [];
  const client = new GideonDBClient("https://db.example", { fetch: queuedFetch([
    jsonResponse({ records: [{ id: "one" }], next_cursor: "next/value", vectors_included: false }),
    jsonResponse({ records: [], next_cursor: "", vectors_included: false, metadata_epoch: 7, authoritative_placement: true }),
  ], requests) });
  const page = await client.scroll("docs", { namespace: "tenant one", limit: 25, cursor: "prior/value", includeVector: true });
  assert.equal(page.next_cursor, "next/value");
  assert.match(requests[0].url, /namespace=tenant\+one/);
  assert.match(requests[0].url, /cursor=prior%2Fvalue/);
  assert.match(requests[0].url, /include_vector=true/);
  const clusterPage = await client.distributedScroll("docs", { limit: 25 });
  assert.equal(clusterPage.metadata_epoch, 7);
  assert.match(requests[1].url, /\/v1\/cluster\/collections\/docs\/vectors\?/);
  await assert.rejects(client.scroll("docs", { limit: 201 }), TypeError);
  await assert.rejects(client.distributedScroll("docs", { limit: 201 }), TypeError);
});

File: pkg/client/client_test.go
package client_test

import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/Bharanipbk/gideondb/internal/api/rest"
	"github.com/Bharanipbk/gideondb/internal/engine"
	"github.com/Bharanipbk/gideondb/pkg/client"
)

type roundTripFunc func(*http.Request) (*http.Response, error)

func (function roundTripFunc) RoundTrip(request *http.Request) (*http.Response, error) {
	return function(request)
}

func inMemoryHTTPClient(handler http.Handler) *http.Client {
	return &http.Client{Transport: roundTripFunc(func(request *http.Request) (*http.Response, error) {
		response := httptest.NewRecorder()
		handler.ServeHTTP(response, request)
		return response.Result(), nil
	})}
}

func TestClientLifecycleAndTypedErrors(t *testing.T) {
	db, err := engine.Open(t.TempDir())
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()
	const apiKey = "0123456789abcdef"
	httpClient := inMemoryHTTPClient(rest.NewWithOptions(db, nil, rest.Options{APIKey: apiKey, NodeID: "11111111111111111111111111111111", ClusterID: "22222222222222222222222222222222", AdvertiseAddress: "gideondb.test:6333"}).Handler())
	sdk, err := client.New("http://gideondb.test", client.Options{APIKey: apiKey, HTTPClient: httpClient})
	if err != nil {
		t.Fatal(err)
	}
	ctx := context.Background()
	if err := sdk.Health(ctx); err != nil {
		t.Fatal(err)
	}
	created, err := sdk.CreateCollection(ctx, client.CollectionConfig{Name: "docs", Dimension: 2, Metric: "dot", ShardCount: 2})
	if err != nil || created.Name != "docs" {
		t.Fatalf("created=%#v err=%v", created, err)
	}
	stored, err := sdk.Upsert(ctx, "docs", client.Record{ID: "one", Namespace: "tenant-a", Vector: []float32{1, 0}, Metadata: map[string]any{"kind": "guide"}})
	if err != nil || stored.Version == 0 {
		t.Fatalf("stored=%#v err=%v", stored, err)
	}
	if _, err := sdk.BatchUpsert(ctx, "docs", []client.Record{{ID: "two", Vector: []float32{0, 1}}}); err != nil {
		t.Fatal(err)
	}
	page, err := sdk.Scroll(ctx, "docs", client.ScrollOptions{Limit: 1})
	if err != nil || len(page.Records) != 1 || page.NextCursor == "" || len(page.Records[0].Vector) != 0 || page.VectorsIncluded {
		t.Fatalf("page=%#v err=%v", page, err)
	}
	next, err := sdk.Scroll(ctx, "docs", client.ScrollOptions{Limit: 2, Cursor: page.NextCursor, IncludeVector: true})
	if err != nil || len(next.Records) != 1 || len(next.Records[0].Vector) != 2 || !next.VectorsIncluded {
		t.Fatalf("next=%#v err=%v", next, err)
	}
	clusterPage, err := sdk.DistributedScroll(ctx, "docs", client.ScrollOptions{Limit: 2})
	if err != nil || len(clusterPage.Records) != 2 || clusterPage.MetadataEpoch == 0 {
		t.Fatalf("cluster page=%#v err=%v", clusterPage, err)
	}
	got, err := sdk.Get(ctx, "docs", "tenant-a", "one")
	if err != nil || got.ID != "one" || got.Namespace != "tenant-a" {
		t.Fatalf("got=%#v err=%v", got, err)
	}
	results, err := sdk.Search(ctx, "docs", client.SearchOptions{Vector: []float32{1, 0}, TopK: 1, Namespace: "tenant-a", Filter: map[string]any{"kind": "guide"}})
	if err != nil || len(results) != 1 || results[0].ID != "one" {
		t.Fatalf("results=%#v err=%v", results, err)
	}
	description, err := sdk.DescribeCollection(ctx, "docs")
	if err != nil || description.VectorCount != 2 {
		t.Fatalf("description=%#v err=%v", description, err)
	}
	collections, err := sdk.ListCollections(ctx)
	if err != nil || len(collections) != 1 {
		t.Fatalf("collections=%#v err=%v", collections, err)
	}
	if err := sdk.Delete(ctx, "docs", "tenant-a", "one"); err != nil {
		t.Fatal(err)
	}
	_, err = sdk.Get(ctx, "docs", "tenant-a", "one")
	var apiErr *client.APIError
	if !errors.As(err, &apiErr) || apiErr.StatusCode != 404 || apiErr.Code != "not_found" {
		t.Fatalf("error=%#v", err)
	}
	if err := sdk.DeleteCollection(ctx, "docs"); err != nil {
		t.Fatal(err)
	}
}

func TestClientScrollRejectsInvalidLimit(t *testing.T) {
	sdk, err := client.New("http://gideondb.test", client.Options{})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := sdk.Scroll(context.Background(), "docs", client.ScrollOptions{Limit: 201}); err == nil {
		t.Fatal("accepted oversized scroll page")
	}
	if _, err := sdk.DistributedScroll(context.Background(), "docs", client.ScrollOptions{Limit: 201}); err == nil {
		t.Fatal("accepted oversized distributed scroll page")
	}
}

func TestClientDistributedEndpointsAndValidation(t *testing.T) {
	db, err := engine.Open(t.TempDir())
	if err != nil {
		t.Fatal(err)
	}
	defer db.Close()
	if _, err := client.New("ftp://invalid", client.Options{}); err == nil {
		t.Fatal("accepted non-HTTP base URL")
	}
	httpClient := inMemoryHTTPClient(rest.NewWithOptions(db, nil, rest.Options{
		NodeID: "11111111111111111111111111111111", ClusterID: "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
		AdvertiseAddress: "node-a:6333", MetadataEpoch: 1,
	}).Handler())
	sdk, err := client.New("http://gideondb.test", client.Options{HTTPClient: httpClient})
	if err != nil {
		t.Fatal(err)
	}
	ctx := context.Background()
	if _, err := sdk.CreateCollection(ctx, client.CollectionConfig{Name: "distributed", Dimension: 2, Metric: "dot", ShardCount: 2}); err != nil {
		t.Fatal(err)
	}
	write, err := sdk.DistributedBatchUpsert(ctx, "distributed", []client.Record{{ID: "one", Vector: []float32{1, 0}}}, "quorum")
	if err != nil || write.Partial || len(write.Outcomes) != 1 {
		t.Fatalf("write=%#v err=%v", write, err)
	}
	search, err := sdk.DistributedSearch(ctx, "distributed", client.SearchOptions{Vector: []float32{1, 0}, TopK: 1})
	if err != nil || search.Partial || len(search.Results) != 1 || search.Results[0].ID != "one" {
		t.Fatalf("search=%#v err=%v", search, err)
	}
}
